segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -331,37 +490,118 @@ private long applyLatestSnapshot(MetadataTransaction txn, HashMap<String, Long>
+                } else if (segmentSnapshot.chunkMetadataCollection.size() == 1) {
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getFirstChunk() != null);
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getFirstChunk().equals(segmentSnapshot.segmentMetadata.getLastChunk()));
+                } else {
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getFirstChunk() != null);
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getLastChunk() != null);
+                    Preconditions.checkState(!segmentSnapshot.segmentMetadata.getFirstChunk().equals(segmentSnapshot.segmentMetadata.getLastChunk()));
+                }
+                ChunkMetadata previous = null;
+                for (val metadata : segmentSnapshot.getChunkMetadataCollection()) {
+                    if (previous != null) {
+                        Preconditions.checkState(previous.getNextChunk().equals(metadata.getName()));
+                    }
+                    previous = metadata;
+                }
+            }
+        }
     }
 
     /**
      * Read contents from file.
      */
-    private byte[] getContents(String snapshotFile) throws ExecutionException, InterruptedException {
-        val info = chunkStorage.getInfo(snapshotFile).get();
-        val h = ChunkHandle.readHandle(snapshotFile);
-        byte[] contents = new byte[Math.toIntExact(info.getLength())];
-        long fromOffset = 0;
-        int remaining = contents.length;
-        while (remaining > 0) {
-            int bytesRead = chunkStorage.read(h, fromOffset, remaining, contents, Math.toIntExact(fromOffset)).get();
-            remaining -= bytesRead;
-            fromOffset += bytesRead;
+    private byte[] getContents(String snapshotFile) throws Exception {
+        Exception e = null;
+        for (int i = 0; i < config.getMaxJournalReadAttempts(); i++) {
+            try {
+                val info = chunkStorage.getInfo(snapshotFile).get();
+                val h = ChunkHandle.readHandle(snapshotFile);
+                byte[] contents = new byte[Math.toIntExact(info.getLength())];
+                long fromOffset = 0;
+                int remaining = contents.length;
+                while (remaining > 0) {
+                    int bytesRead = chunkStorage.read(h, fromOffset, remaining, contents, Math.toIntExact(fromOffset)).get();
+                    remaining -= bytesRead;
+                    fromOffset += bytesRead;
+                }
+                return contents;
+            } catch (Exception ex) {
+                e = ex;
+                log.warn("SystemJournal[{}] Error while reading journal {}.", containerId, snapshotFile, e);
+            }
+        }
+        if (null != e) {
+            throw e;
         }
-        return contents;
+        return new byte[0];
```

Comment:
Why would we get here? Is it because we found no journal?

If so, please log a WARN saying so, and also explaining something like "This is OK if it's the first time initializing the container" (or whatever message is appropriate for this case).

Reply:
Can happen when file is empty.

Reply:
Added warning.